### PR TITLE
Create ec2_asg_scheduled_action module

### PR DESCRIPTION
**Pull request by marknet15**
##### SUMMARY
This creates a new module `ec2_asg_scheduled_action` to create scheduled actions on Auto Scaling Groups.

It was based on and modified from: https://github.com/mmochan/ansible-aws-ec2-asg-scheduled-actions/blob/master/library/ec2_asg_scheduled_action.py

Requires: https://github.com/mattclay/aws-terminator/pull/179

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
ec2_asg_scheduled_action

##### ADDITIONAL INFORMATION
Actions can be created like so:
```yml
- name: Create a minimal scheduled action for autoscaling group
  community.aws.ec2_asg_scheduled_action:
    autoscaling_group_name: test_asg
    scheduled_action_name: test_scheduled_action
    start_time: 2021 October 25 08:00 UTC
    recurrence: 40 22 * * 1-5
    desired_capacity: 10
    state: present
```

Actions can be deleted like so:
```yml
- name: Delete scheduled action
  community.aws.ec2_asg_scheduled_action:
    autoscaling_group_name: test_asg
    scheduled_action_name: test_scheduled_action
    state: absent
```

### CODE REVIEW

**Comment by marknet15** [comment]
I've created https://github.com/mattclay/aws-terminator/pull/179 that is required for the tests

File: plugins/modules/ec2_asg_scheduled_action.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    if not HAS_DATEUTIL:
+        module.fail_json(msg='dateutil is required for this module')
+
+    client = module.client('autoscaling', retry_decorator=AWSRetry.jittered_backoff())
+    current_actions = get_scheduled_actions()
+    state = module.params.get('state')
+    results = dict()
+
+    if state == 'present':
+        changed = put_scheduled_update_group_action(current_actions)
```
Comment (markuman) [inline plugins/modules/ec2_asg_scheduled_action.py:308]: This is not covered by `check_mode`.

Comment (marknet15) [reply plugins/modules/ec2_asg_scheduled_action.py:308]: @markuman so it calls the function and has a check for check mode here:
https://github.com/ansible-collections/community.aws/pull/779/files#diff-9af6ca29f1ac5f35e2679d6d8eadc230ac054671a271f2d50d8ce9e0898a3866R265

File: tests/integration/targets/ec2_asg_scheduled_action/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        desired_capacity: 2
+        state: present
+      register: scheduled_action
+
+    - name: Check results - Create basic scheduled_action
+      assert:
+        that:
+          - scheduled_action is successful
+          - scheduled_action is changed
+          - scheduled_action.autoscaling_group_name == "{{ resource_prefix }}-asg"
```
Comment (jillr) [inline tests/integration/targets/ec2_asg_scheduled_action/tasks/main.yml:115]: `autoscaling_group_name` is not part of the results dict

File: plugins/modules/ec2_asg_scheduled_action.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        params['DesiredCapacity'] = module.params.get('desired_capacity')
+
+    if module.params.get('min_size') is not None:
+        params['MinSize'] = module.params.get('min_size')
+
+    if module.params.get('max_size') is not None:
+        params['MaxSize'] = module.params.get('max_size')
+
+    if module.params.get('time_zone') is not None:
+        params['TimeZone'] = module.params.get('time_zone')
```
Comment (jillr) [inline plugins/modules/ec2_asg_scheduled_action.py:197]: This fails for me:
```
Unknown parameter in input: "TimeZone", must be one of: AutoScalingGroupName, ScheduledActionName, Time, StartTime, EndTime, Recurrence, MinSize, MaxSize, DesiredCapacity
```
Though I see it in the boto docs.  Is a newer version of botocore required for this module than the collection minimum? https://github.com/ansible-collections/community.aws/blob/main/requirements.txt  
If it is then we need to document that in the module docs for the affected parameter and do a check for it, like in this example https://github.com/ansible-collections/community.aws/blob/main/plugins/modules/aws_msk_cluster.py#L580

**Review by jillr — COMMENTED** [review]
Couple things I found while testing the IAM policies

File: plugins/modules/ec2_asg_scheduled_action.py (hunk 3 above)
Comment (marknet15) [reply plugins/modules/ec2_asg_scheduled_action.py:197]: Yes I think it's quite a recent addition, Makes sense I'll add the check. Thanks for the pointer 🙏

File: plugins/modules/ec2_asg_scheduled_action.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+    if len(current_actions) < 1:
+        changed = True
+    else:
+        for k, v in params.items():
+            # To correctly detect changes convert the start_time & end_time to datetime object
+            if k == "StartTime":
+                v = timedate_parse(v)
+            elif k == "EndTime":
+                v = timedate_parse(v)
```
Comment (tremble) [inline plugins/modules/ec2_asg_scheduled_action.py:257]: There's no need to loop to find the parameters 

```suggestion
        # To correctly detect changes convert the start_time & end_time to datetime object
        if "StartTime" in params:
            params["StartTime"] = timedate_parse(params["StartTime"])
        if "EndTime" in params:
            params["EndTime"] = timedate_parse(params["EndTime"])
        for k, v in params.items():
```

File: plugins/modules/ec2_asg_scheduled_action.py (hunk 3 above)
Comment (tremble) [reply plugins/modules/ec2_asg_scheduled_action.py:197]: Added in botocore 1.20.24  (Tue Mar 9 19:45:18 2021)

File: plugins/modules/ec2_asg_scheduled_action.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+# Based off of https://github.com/mmochan/ansible-aws-ec2-asg-scheduled-actions/blob/master/library/ec2_asg_scheduled_action.py
+# (c) 2016, Mike Mochan <@mmochan>
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: ec2_asg_scheduled_action
+version_added: 2.1.0
```
Comment (markuman) [inline plugins/modules/ec2_asg_scheduled_action.py:15]:
```suggestion
version_added: 3.0.0
```

**Review by markuman — APPROVED** [review]
just the version_added. I've committed it.

**Review by alinabuzachis — APPROVED** [review]
@marknet15 Thank you for working on this. It looks good to me!

@markuman If we want to have this included in the 2.2.0 release, we have to update `version_added`.

File: plugins/modules/ec2_asg_scheduled_action.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+# Based off of https://github.com/mmochan/ansible-aws-ec2-asg-scheduled-actions/blob/master/library/ec2_asg_scheduled_action.py
+# (c) 2016, Mike Mochan <@mmochan>
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: ec2_asg_scheduled_action
+version_added: 3.0.0
```
Comment (markuman) [inline plugins/modules/ec2_asg_scheduled_action.py:15]:
```suggestion
version_added: 2.2.0
```

**Comment by markuman** [comment]
> @marknet15 Thank you for working on this. It looks good to me!
> 
> @markuman If we want to have this included in the 2.2.0 release, we have to update `version_added`.

Ok, once the ci is green again, I'll attache `backport-2` and `gate` label.

**Pull request merged**